### ISSUE TITLE
Nodes: Optimization: Destructuring with non-final spread should still use rest syntax

### ISSUE THREAD

**Issue by isiahmeadows** [jashkenas/coffeescript#4517]
Example:

```coffee
tag = "<impossible>"

[open, contents..., close] = tag.split("")
```

```js
// Recommended output:
var close, contents, open, tag, pop = [].pop;

tag = "<impossible>";

[open, ...contents] = tag.split("");
close = pop.call(contents);
```

**Comment by zdenko** [comment]
This definitely looks cleaner.
But, I think, using `array.pop` will work only until the number of destructured variables (without the splat) is the same as the length of the array.

Example:

```
# CS:
[a, rest…, b, c, d] = ["a", "b", "c", "d"]

# JS:
[a, …rest] = ["a", "b", "c", "d"]
d = pop.call(rest)
c = pop.call(rest)
b = pop.call(rest)

# Result:
a = "a"
rest = []
d = "d"
c = "c"
b = "b"
```

However, in this case the result will not be as expected:
```
# CS:
[a, rest…, b, c, d] = ["a", "b"]

# JS
…

# Result:
a = "a"
rest = []
d = "b"
c = undefined
b = undefined
```

Perhaps using something like `array.splice` would be a solution:

```
# CS: 
# splice = [].splice
# [a, rest…, b, c, d] = ["a", "b"]

# JS:
[a, …rest] = ["a", "b"]
[b, c, d] = (3 <= rest.length) ? splice.call(rest, rest.length - 3, rest.length) : splice.call(rest, 0);

# Result:
a = "a"
rest = []
b = "b"
c = undefined
d = undefined
```

**Comment by isiahmeadows** [comment]
Maybe, this, then?

```js
# [a, rest…, b, c, d] = ["a", "b"]

[a, …rest] = ["a", "b", "c", "d"];
_i = rest.length - 3; if (i < 0) i = 0;
if (_i !== rest.length) b = rest[_i++];
if (_i !== rest.length) c = rest[_i++];
if (_i !== rest.length) d = rest[_i++];
rest.length = _i;
```

That should avoid the issue without resorting to a method call.

**Comment by zdenko** [comment]
Method calls are not the problem. It's just that `array.pop` method removes the **last** element from an array.

Your last solution would also work. There are few points which need correction:
1. A small typo: `_i` instead `i`
```coffeescript
 _i = rest.length - 3; if (_i < 0) _i = 0; 
```
Which can be also shortened:
```coffeescript
_i = 3 <= rest.length ? rest.length - 3 : 0
```

```coffeescript
# [a, ...rest, b, c, d] = ["a", "b", "c", "d"]
[a, ...rest] = ["a", "b", "c", "d"]
_i = 3 <= rest.length ? rest.length - 3 : 0
# Result:
# a = "a"
# rest = ["b", "c", "d"]
# _i = 0
```
2. No need for checking: `_i !== rest.length` 
```coffeescript
b = rest[_i++]  # b = "b", _i = 1
c = rest[_i++]  # c = "c", _i = 2
d = rest[_i++]  # d = "d", _i = 3
```
3. Since `_i` is now 3, `rest` will not change
```coffeescript
rest.length = _i
# rest = ["b", "c", "d"]
```

So, something like this will work:
```coffeescript
rest.length = 3 <= _i ? _i - 3 : 0
# rest = []
```

But, even in this case, one could prevent changing the property `length`, and `rest` will not be changed:
```coffeescript
# rest = ["b", "c", "d"]
Object.defineProperty(rest, "length", {writable: false})
rest.length = 0
# rest = ["b", "c", "d"]
```

**Comment by isiahmeadows** [comment]
@zdenko 

I'm addressing each one out of order:

> Method calls are not the problem. It's just that array.pop method removes the last element from an array.

I'm aware. I simply decided to avoid it in this go around.

> A small typo: _i instead i

Thanks! I updated my comment accordingly.

> Which can be also shortened:

Your version is longer:

```js
// Expanded
_i = rest.length - 3; if (_i < 0) _i = 0;
_i = 3 <= rest.length ? rest.length - 3 : 0;

// Minified
i=r.length-3;if(i<0)i=0;
i=3<=r.length?r.length-3:0;
```

> Since _i is now 3, rest will not change

My case uses conditionals to avoid that mess.

> No need for checking: _i !== rest.length
>
> ```js
> b = rest[_i++]  # b = "b", _i = 1
> c = rest[_i++]  # c = "c", _i = 2
> d = rest[_i++]  # d = "d", _i = 3
> ```
>
> But, even in this case, one could prevent changing the property length, and rest will not be changed: [...]

Actually, it wouldn't work that way, because they wouldn't be able to read `...rest` unless they have defaulted values, and that issue could be solved by computing them *after* the truncation and destructuring.

You *could* make the argument of `Array.prototype.length` being overwritten, but I could make the same argument about you accessing out of bounds indices, by defining a setter for `Array.prototype[2]` (which gives me access to the `rest` prematurely).

**Comment by zdenko** [comment]
@isiahmeadows point taken.
Indeed, my version is longer. 

I'm not sure about the conditionals.
```coffeescript
# [a, rest..., b, c, d] = ["a", "b", "c", "d", "e", "f", "g", "h"]
[a, ...rest] = ["a", "b", "c", "d", "e", "f", "g", "h"]
# a = "a"
# rest = ["b", "c", "d", "e", "f", "g", "h"]
_i = rest.length - 3; if (_i < 0) _i = 0
# _i = 7 - 3 = 4
if (_i !== rest.length) b = rest[_i++]
# b = "f", _i = 4++ = 5
if (_i !== rest.length) c = rest[_i++]
# c = "g", _i = 5++ = 6
if (_i !== rest.length) d = rest[_i++]
# d = "h", _i = 6++ = 7
rest.length = _i
# rest.length = 7, rest = ["b", "c", "d", "e", "f", "g", "h"]
```

So, perhaps:
```coffeescript
rest.length = _i - 3
# rest.length = 7 - 3 = 4, rest = ["b", "c", "d", "e"]
```

But, in this case:
```coffeescript
[a, rest..., b, c, d] = ["a", "b"]
...
# a = "a"
# b = "b"
# _i = 1
# rest.length = _i - 3 = 1 - 3 => RangeError
```

The same example without conditional `_i !== rest.length` would work:
```coffeescript
[a, ...rest] = ["a", "b"]
# a = "a"
# rest = ["b"]
_i = rest.length - 3; if (_i < 0) _i = 0
# _i = 2 - 3 = 0
b = rest[_i++]
# b = "b", _i = 0++ = 1
c = rest[_i++]
# c = undefined, _i = 1++ = 2
d = rest[_i++]
# d = undefined, _i = 2++ = 3
rest.length = _i - 3
# rest.length = 0, rest = []
```

Anyway, to start at the beginning, you made good suggestions for an alternative approach which avoids the need for additional  `ref` variable in `[a, rest..., b] = [1, 2, 3, 4]`, and (re)use `rest` to get values.

**Comment by isiahmeadows** [comment]
I see what you mean about the conditionals (it'll be an out of bounds access, but the engine should generate similar code either way).

> Anyway, to start at the beginning, you made good suggestions for an alternative approach which avoids the need for additional ref variable in [a, rest..., b] = [1, 2, 3, 4], and (re)use rest to get values.

Yep, and thanks! The whole reason I recommended it was to avoid generating extra garbage for an inaccessible temp variable.

**Comment by GeoffreyBooth** [comment]
Implemented in #4825.
